### PR TITLE
[Performance] V1 Pooling Models E2E Performance Optimization

### PR DESCRIPTION
**Pull request by noooop**
## Purpose

V1 Pooling Models E2E Performance Optimization

```
#  https://github.com/noooop/snippet/blob/main/benchmarks/embed/profile.py
VLLM_TORCH_PROFILER_DIR=/xxx python profile.py
```

main CPU: 67.134ms CUDA: 46.417ms -> this pr  CPU: 39.740ms CUDA: 39.890ms

benchmarks: 
- sentence_transformers: https://github.com/noooop/snippet/blob/main/benchmarks/embed/st.py
- v0: https://github.com/noooop/snippet/blob/main/benchmarks/embed/v0.py
- v1: https://github.com/noooop/snippet/blob/main/benchmarks/embed/v1.py

Result:
- X-axis: Throughput (token/s)
- Y-axis: Latency, Time needed for one step (ms)
- The curve lower right is better ↘ 

image

image

image

image

image

Frankly speaking,
this optimizations only show significant improvements when handling many small requests—they have almost no impact on large requests (since the backbone network latency dominates in such cases).

Of course, a faster implementation is always cooler.

Please click here for long details.
↓↓↓↓↓↓↓↓↓
<details>

1. replace_roberta_positions takes too long

image

6d8f55e: CPU: 67.134ms CUDA: 46.417ms -> CPU: 53.261ms CUDA: 42.820ms

Correct Testing
- pytest -s -vvv tests/models/language/pooling/test_baai.py
   - BAAI/bge-m3 for XLMRobertaModel
   - BAAI/bge-reranker-base for XLMRobertaForSequenceClassification

2. reduce cuda sync
image

8ff2418: CPU: 53.261ms CUDA: 42.820ms -> CPU: 43.016ms CUDA: 42.516ms

3. Use as many batch operations as possible

image

13e44df: CPU: 43.016ms CUDA: 42.516ms -> CPU: 42.463ms CUDA: 40.629ms

4. non_blocking seq_lens (Use seq_lens_cpu directly later

image

876cb9a: CPU: 42.463ms CUDA: 40.629ms -> CPU: 39.934ms CUDA: 40.128ms

5. remove prompt_len == hidden_states.shape[0] cuda sync by using prompt_len (cpu tenser) == num_scheduled_tokens (cpu tenser)

image

93d0a95: CPU: 39.934ms CUDA: 40.128ms -> CPU: 39.509ms CUDA: 40.094ms

6. _decode_token_type_ids uses ones_like to avoid getting shape which causes cuda sync

image

f649899: CPU: 39.509ms CUDA: 40.094ms -> CPU: 39.317ms CUDA: 40.109ms

7. non_blocking torch.zeros in _build_encoder_only_attn_metadata

image

d7fa9a8: CPU: 39.317ms CUDA: 40.109ms -> CPU: 39.182ms CUDA: 39.981ms

8. using pooling_cursor Rather than torch.split(hidden_states, num_scheduled_tokens_list)

image

2f10806: CPU: 39.182ms CUDA: 39.981ms -> CPU: 39.740ms CUDA: 39.890ms

</details>

## Test Plan

keep CI green

## (Optional) Documentation Update

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/models/roberta.py [hunk 1]
```diff
@@ -233,58 +232,14 @@ def forward(
...
-                create_position_ids_from_input_ids(tokens, padding_idx)
-            offset = offset + length
+    # Replace position ids because in RoBERTa models
+    # they have to start at padding_idx + 1 and ignore
+    # existing padding tokens
+    # References:
+    # - https://github.com/huggingface/transformers/blob/a3d69a8994d673899608a7c17fbf4f953f50474e/src/transformers/models/roberta/modeling_roberta.py#L133
+    # - https://github.com/huggingface/transformers/blob/a3d69a8994d673899608a7c17fbf4f953f50474e/src/transformers/models/roberta/modeling_roberta.py#L1669
+    # vllm does not use padding tokens, let's make things simpler
+    position_ids += padding_idx + 1
```
Comment (noooop) [inline vllm/model_executor/models/roberta.py:245]: @maxdebayser 

Performance optimization while ensuring correct is very tricky.

Does this modification work for you?

Comment (maxdebayser) [reply vllm/model_executor/models/roberta.py:245]: @noooop , if none of the `input_ids` is `padding_idx` I think your code is 100% correct. How much performance impact would the be if you did something like this?

```
# Pseudo code:
if padding_idx in input_ids:
    run_old_code()
 else:
     position_ids += padding_idx + 1
```

Currently there are only two situations where `padding_idx` could show up in the input_ids:
1) For cross-encoder models as a separator
2) If the user sends them as part of the input. I don't know why they would do that, but technically it's possible. We could make it unsupported though.

There is also a separate conversation going on about using torch sdpa or some other attention backend directly in the modeling code to simplify the gpu model runner and the code in `v1/attention`. In this case I don't know if we'll have to use padded batches. However, in this case the padding tokens would be masked out anyway, so it's probably not a problem to optimize.

Comment (noooop) [reply vllm/model_executor/models/roberta.py:245]: Any branch here will call cuda stream sync, leading to very slow performance

Comment (noooop) [reply vllm/model_executor/models/roberta.py:245]: https://huggingface.co/BAAI/bge-reranker-base/blob/main/special_tokens_map.json
https://huggingface.co/BAAI/bge-reranker-v2-m3/blob/main/special_tokens_map.json

```
"pad_token": "<pad>"
"sep_token": "</s>"
```

At least pad_token and sep_token are different

> However, in this case the padding tokens would be masked out anyway, so it's probably not a problem to optimize.  +1

I can't think of any scenario where vllm using pad_token. 

I also can't imagine a scenario where the user input <pad> would produce meaningful output, since we perform attention padding based on attention metadata, not based on the token itself.

Comment (maxdebayser) [reply vllm/model_executor/models/roberta.py:245]: > At least pad_token and sep_token are different

Nice! So if all classification tests pass I think we can keep the optimization

> I can't think of any scenario where vllm using pad_token.

> I also can't imagine a scenario where the user input would produce meaningful output, since we perform  attention padding based on attention metadata, not based on the token itself.

If we move to using an attention implementation directly in the model code then there could be an impact because there will be no attention metadata. And depending on the attention backend we might have to run a padded batch instead of a concatenated one. 

With flash_attn in vLLM a batch for 3 sequences looks like this: `[s1,s2, s2,s2,s2,s4,s3,s3,s3]` but other attention backends such as torch sdp require a batch dimension, so it could be:
```
[[s1, s1,  P,  P],
 [s2, s2, s2, s2],
 [s3, s3, s3,  P]]
 ```

Comment (noooop) [reply vllm/model_executor/models/roberta.py:245]: If v1 adds an impl of torch sdpa attention backends, it will also use a loop to run forward once for each seq, instead of using pad_token, just like the v0 implementation. 

Otherwise, it would not be compatible with other attention backends implementations.

https://github.com/vllm-project/vllm/blob/a5dd03c1ebc5e4f56f3c9d3dc0436e9c582c978f/vllm/attention/backends/torch_sdpa.py#L646-L659

File: vllm/v1/pool/metadata.py [hunk 2]
```diff
@@ -8,18 +8,66 @@
...
         )
+
+    def build_pooling_cursor(self, num_scheduled_tokens: list[int],
+                             device: torch.device):
+        self.pooling_cursor = build_pooling_cursor(num_scheduled_tokens,
+                                                   self.prompt_lens.tolist(),
+                                                   device)
+
+
+def build_pooling_cursor(num_scheduled_tokens: list[int],
```
Comment (maxdebayser) [inline vllm/v1/pool/metadata.py:60]: Is there a reason to convert the tensors to list as arguments for this function and then convert to tensor again?

File: vllm/model_executor/models/roberta.py (hunk 1 above)
Comment (maxdebayser) [reply vllm/model_executor/models/roberta.py:245]: > If v1 adds an impl of torch sdpa attention backends, it will also use a loop to run forward once for each seq, instead of using pad_token, just like the v0 implementation.

That's an option, yes. But what if it's faster to copy to a [B,s,h] tensor we might need padding. The nice thing about all this benchmark work we're doing is that we'll be able to test. Anyway, you've already shown that the separator token is not the padding token in all models that we know about, so there is no impact for this PR.

Comment (maxdebayser) [reply vllm/model_executor/models/roberta.py:245]: But thanks for pointing me to this code, I wasn't aware of this implementation choice in V0

File: vllm/v1/pool/metadata.py (hunk 2 above)
Comment (noooop) [reply vllm/v1/pool/metadata.py:60]: For compatibility with v0

**Comment by maxdebayser** [comment]
@noooop , nice work! There are two optimizations in this PR: the roberta position ids and removing the split of the hidden states. Do you have benchmark results on how much each individually changes the performance compared to the `main` branch?

File: vllm/v1/pool/metadata.py (hunk 2 above)
Comment (noooop) [reply vllm/v1/pool/metadata.py:60]: This pooling_cursor doesn’t show almost any performance improvement compared to torch.split, except for being cooler.

LOL

Comment (noooop) [reply vllm/v1/pool/metadata.py:60]: I guess torch.split doesn’t actually copy any data under the hood.

Actually, it also uses some kind of cursor internally.

**Comment by noooop** [comment]
> @noooop , nice work! There are two optimizations in this PR: the roberta position ids and removing the split of the hidden states. Do you have benchmark results on how much each individually changes the performance compared to the `main` branch?

Please click detail for long details.

**Comment by noooop** [comment]
these optimizations only show significant improvements when handling many small requests—they have almost no impact on large requests (since the backbone network latency dominates in such cases).

File: vllm/v1/pool/metadata.py (hunk 2 above)
Comment (maxdebayser) [reply vllm/v1/pool/metadata.py:60]: Yeah, when I implemented the split in the gpu model runner, I thought that the split itself would be basically free because it doesn't have to copy any data.  But I suspected that running loops on the split tensors could potentially be slower than running operations on the full batch. The reason I still did it despite the possible performance downside was because of the idea of allowing pooling plugins in the future. With list of tensors it's potentially easier to write and chain these plugins. But I like your idea of cursors, I think it can be evolved in the future if we implement plugins.

Comment (maxdebayser) [reply vllm/v1/pool/metadata.py:60]: Actually, it just occurred to me that in the vllm-spyre backend the model output is [B,s,h]: https://github.com/vllm-project/vllm-spyre/blob/d0dd0e1ec4eda0bcc6bea4b4840513e30bb9cbda/vllm_spyre/v1/worker/spyre_model_runner.py#L1498 . For LAST, CLS and ALL pooling, I think I can just build the cursor with 2D indices to select the correct positions. But I don't know if the MEAN implementation would work out of the box. What are your thoughts on this? Do you mind if I test this before this PR is merged?

**Comment by noooop** [comment]
@DarkLight1337

Ready for review

File: vllm/v1/pool/metadata.py [hunk 3]
```diff
@@ -6,20 +6,74 @@
...
+pin_memory = is_pin_memory_available()
+
+
+@dataclass
+class PoolingCursor:
+    index: list[int]
+    first: torch.Tensor  # GPU Tensor
+    last: torch.Tensor  # GPU Tensor
+    prompt_lens: torch.Tensor  # CPU Tensor
+    num_scheduled_tokens: torch.Tensor  # CPU Tensor
```
Comment (DarkLight1337) [inline vllm/v1/pool/metadata.py]:
```suggestion
    first_token_indices_gpu: torch.Tensor
    last_token_indices_gpu: torch.Tensor
    prompt_lens_cpu: torch.Tensor
    num_scheduled_tokens_cpu: torch.Tensor
```

Improve the naming

File: vllm/v1/pool/metadata.py (hunk 2 above)
Comment (noooop) [reply vllm/v1/pool/metadata.py:60]: I'm glad to verify, but I need a bit of time to figure out how to install vllm-spyre

**Comment by noooop** [comment]
@DarkLight1337 

Please unblock `Language Models Test (Extended Pooling)` to make sure I didn't break anything

**Comment by noooop** [comment]
@DarkLight1337 

By the way, since #22878 Pooling models mteb test uses enforce_eager, mteb_test_embed_models has become less (never) flaky.

 I'll reset the threshold back to MTEB_EMBED_TOL = 1e-4 next month.

image

I'm not 100% sure that flaky is caused by torch.compile. It is very likely caused by it.

cc @robertgshaw2-redhat

**Review by DarkLight1337 — APPROVED** [review]
Tests pass so this should be good to go, thanks!
